### ISSUE TITLE
Discussion: How best to add cold gas props.

### ISSUE THREAD

**Issue by MatthieuSchaller** [ICRAR/VELOCIraptor-STF#67]
I am looking at adding H_2 and HI masses in the catalogs, and especially to the aperture measurements. 

Some advice would be nice here. 

I could:
 - Add a H_2 mass to the Particle object and just fill it in when reading in the snapshot. This is made of a combination of snapshot fields (mass * H_frac * H2_frac * 2).
  - Add a species map<> to the HydroProperties in the Particle object and store in there the H2_frac. Then when summing the mass of H2 for a given structure compute H2_mass from the species and already existing mass and H element fraction.

In spirit, it's similar to the SFR we accumulate but is more specific to SWIFT-EAGLE++.

@pelahi any thoughts on what the best choice would be to keep things tidy before I start typing?

**Comment by pelahi** [comment]
HI @MatthieuSchaller , I will write up some guidance soon. I just wanted to let you know that I will be off from work tomorrow and the following monday but still should get a chance to write something.

**Comment by rtobar** [comment]
FWIW I have a set of changes that I haven't pushed yet to simplify the declaration of the output datasets written by VR, so this could be a good opportunity to include those changes in.

**Comment by MatthieuSchaller** [comment]
Thanks both. Just some basic style recommendations would suffice; I can likely take it from there.

**Comment by MatthieuSchaller** [comment]
Is there any news on this? Any quick advice at this stage would suffice to get started and draft some code.

**Comment by pelahi** [comment]
Hi @MatthieuSchaller 
Sorry for the late reply, been a bit snowed under with work and family. 
So to be clear, you'd like to add extra Hydrogen properties (namely H2 and HI) to particles and have stuff naturally calculated. But the H_2_mass of gas particles is simply mass *H_frac *H2_frac *2 and HI being mass * H_frac *HI_frac and so on so it does not make sense for swift to store these masses separately. Instead you store just the frac, correct? 

If that's the case (and particularly if you see the fraction being useful on their own), I would add a name convention where gas particles have the frac fields loaded by default. This would be like adding SFR to the gas particles and would require updating hdfitems.h. 

Having done this, the next step is trickier as it requires updating hdfio.cxx to load these fields if the appropriate input is loaded. And also would require then updating the substructure.cxx file to calculate properties based on these quantities if present. 

The other options is to load the fields as you would just any old extra property but specify a specific calculation flag that would then calculate the results you want. I will generate a draft of the later (and will do the same for the former if I have time).

**Comment by MatthieuSchaller** [comment]
Thanks for taking the time. 

The frac themselves are not interesting anymore and definitely not practical when we consider aperture quantities for instance. So, yes, I'd like to carry an HI and H_2 mass for the VR-particles. These happen to be constructed from a combination of snapshot fields as you described. 

Once they are a particle-carried quantity, it becomes "easy" to have HI and H_2 in apertures straight in the catalog.

**Comment by pelahi** [comment]
Hi @MatthieuSchaller so if you are able to just produce the extra input fields then it is a simple matter of adding the appropriate config options so that aperture quantities are calculated. These would be 3D apertures based on all other aperture calculations. Note that it would be necessary to add some routines to get projected aperture quantities for extra fields. The sample config options would be something like 
```
#this is an example of how we use extra properties
#list of extra properties load as an internal property. This is a set of strings separated by ,
#Here is an example of loading several properties as internal properties
#And calculate mass weighted averages
#The code also should be given a conversion factor
#and the units
Gas_internal_property_names=Pressures,Densities,H2_mass,HI_mass,
Gas_internal_property_input_output_unit_conversion_factors=1.0,1.0,1.0,1.0,
Gas_internal_property_calculation_type=averagemassweighted,averagemassweighted,aperture_total,aperture_total,
#here we set the units of pressure (mass *length^-1 * time^-2 ) and density (mass / length^3)
Gas_internal_property_output_units=1:-1:0:-2:,1:-3:0:0:,1:0:0:,1:0:0,
```

I think all the code is in place. I'll have a test (do you have a small snapshot that would be a useful test?). Do you want projected aperture values to?

**Comment by MatthieuSchaller** [comment]
We don't have HI or H_2 masses directly. But I could pre-process the snapshots to construct them and then use the generalised mechanism you suggest, yes. 
Maybe that's the easiest thing to do to start with.

**Comment by MatthieuSchaller** [comment]
Projected apertures are not needed, no.

**Comment by MatthieuSchaller** [comment]
That would all work provided #72 is addressed.

**Comment by MatthieuSchaller** [comment]
This is now all unnecessary as @pelahi's much better solution was used and the required fix to #72 was done. 

Thanks!
